### DISCUSSION TITLE
Nested Loops

### DISCUSSION THREAD

**Discussion by thebruuu** [vega/vega#3970] [Q&A]
Hello

I still have issues on getting the execution logic/philisopy of Vega.
I am facing a rather simple problem
I have some data which have the following structure:
Category family, Item ID
both columns contains string 

I want to make a visualisation whose main mark group iterates over categories (distinc values) and then the nested mark iterate over item ID using the same X for each category family ...
I am lost ....
if it would be some programming language it would be
for each category 
        DO ...
        for each item ID in the category DO ...

**Comment by hydrosquall** [comment]
Hi @thebruuu , I'm not sure I understand the question, so I'll share 2 examples of  graphs that iterate over 2 layers of categories. 

In Vega rather than writing a loop (imperative approach), you describe what happens to each row of data (declarative approach). In this way, working with Vega is more like writing SQL than writing a custom Javascript function.

To show what this means with 2 examples:

- https://vega.github.io/vega-lite/examples/bar_grouped.html  gives 1 group the x channel, another group the xOffset channel.
- Alternately with facet plots ( https://vega.github.io/vega-lite/examples/trellis_stacked_bar.html ) you can give one category a faceting encoding (like row/group/column), and give your other categories a different channel like x/y/color.

What type of visualization are you trying to create with your category family + id columns?
